### PR TITLE
Fix sampler as parameter bug

### PR DESCRIPTION
**Pull request by Jjagg**
This fixes the parse error that would occur when you have a sampler as a parameter in a function described in #5315. Does not yet fix the bug when declaring a sampler as a local parameter.

When a sampler is declared and not immediately assigned inside a function, the parser will not crash but the sampler will be incorrectly registered as a sampler of the shader. When a sampler is declared and assigned in one statement inside a function the parser will crash. I didn't find a good way to solve this yet :/ 

@tomspilman Am I using a different version of TinyPG or something? Asking because of all the missing comments. I got the latest from the SickheadGames branch.

@Danthekilla Can you test your shaders with this branch? Preferably all those that you ported to make sure nothing's broken. You can get the build artifacts from the teamcity page when it's done building. I'll post a link when it's done.

### CODE REVIEW

**Comment by Jjagg** [comment]
@Danthekilla If you're on Windows (which you need to be to build effects) you can get the installer on this page: http://teamcity.monogame.net/viewLog.html?buildId=29435&buildTypeId=MonoGame_PackagingWindows&tab=artifacts
If you prefer not to override your MG installation you can pull this branch into your fork by adding my remote as an upstream. Then you can build the Windows solution and run the pipeline tool exe from the bin folder (or just run the pipeline tool project from VS). If you prefer this, but don't know exactly how send me a message on gitter and I'll help you out.

File: Tools/2MGFX/Scanner.cs [hunk 1]
```diff
@@ -19,22 +19,21 @@ public partial class Scanner
...
-        public Scanner()
+        static Scanner()
         {
             Regex regex;
             Patterns = new Dictionary<TokenType, Regex>();
             Tokens = new List<TokenType>();
-            LookAheadToken = null;
-            Skipped = new List<Token>();
-
+
```
Comment (tomspilman) [inline Tools/2MGFX/Scanner.cs]: This looks like maybe you're using a different version of TinyPG than normal.  Let me investigate.

**Comment by Danthekilla** [comment]
I will test this with all my shades in an hour or two and get back to you with the results. Am I correct in that this will not compile my fxaa effect just yet?

**Comment by Jjagg** [comment]
> Am I correct in that this will not compile my fxaa effect just yet?

It should work! :)

**Comment by Danthekilla** [comment]
All 41 of my custom shaders now build. I will need to do some testing to make sure they are all still looking correct but they compile at least 👍 

Thanks!

**Comment by Danthekilla** [comment]
Some of my other content (models) with a custom processor have broken and now throw strange errors. I am looking into it, I don't think it has anything to do with your changes.

**Comment by Danthekilla** [comment]
This has broken all my model processors when they try to process internal effects from what I can tell.

I have spent the last 5 hours trying to get a source version of the pipeline building (I also tried using your samplerparse branch) and while I can get it to compile and run it fails to open any mgcb projects due to a reflection error on the built in processors...

This is the error I am getting on my content

`D:/Desktop/2.0 Projects/Games/Block Games/Craft World/Source Content/Models/Creatures/Bear_Trap/Bear_Traplow.FBX: error: Processor 'SkinnedModelProcessor' had unexpected failure!
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Xna.Framework.Content.Pipeline.Processors.EffectProcessor.Process(EffectContent input, ContentProcessorContext context)
   at PipelineExtensionsAnimation.SkinnedModelMaterialProcessor.ProcessMaterial(MaterialContent input, MaterialContent output, ContentProcessorContext context) in D:\Desktop\2.0 Projects\Games\Block Games\2.0 Pipeline Extentions\Animation\Pipeline\SkinnedModelMaterialProcessor.cs:line 112
   at PipelineExtensionsAnimation.SkinnedModelMaterialProcessor.Process(MaterialContent input, ContentProcessorContext context) in D:\Desktop\2.0 Projects\Games\Block Games\2.0 Pipeline Extentions\Animation\Pipeline\SkinnedModelMaterialProcessor.cs:line 41
   at Microsoft.Xna.Framework.Content.Pipeline.ContentProcessor`2.Microsoft.Xna.Framework.Content.Pipeline.IContentProcessor.Process(Object input, ContentProcessorContext context)
   at MonoGame.Framework.Content.Pipeline.Builder.PipelineProcessorContext.Convert[TInput,TOutput](TInput input, String processorName, OpaqueDataDictionary processorParameters)
   at PipelineExtensionsAnimation.SkinnedModelProcessor.ProcessMaterial(MaterialContent materialContent, ContentProcessorContext context) in D:\Desktop\2.0 Projects\Games\Block Games\2.0 Pipeline Extentions\Animation\Pipeline\SkinnedModelProcessor.cs:line 673
   at PipelineExtensionsAnimation.SkinnedModelProcessor.ProcessMaterials(NodeContent node, Dictionary`2 processedMaterials, ContentProcessorContext context) in D:\Desktop\2.0 Projects\Games\Block Games\2.0 Pipeline Extentions\Animation\Pipeline\SkinnedModelProcessor.cs:line 613
   at PipelineExtensionsAnimation.SkinnedModelProcessor.ProcessMaterials(NodeContent node, Dictionary`2 processedMaterials, ContentProcessorContext context) in D:\Desktop\2.0 Projects\Games\Block Games\2.0 Pipeline Extentions\Animation\Pipeline\SkinnedModelProcessor.cs:line 625
   at PipelineExtensionsAnimation.SkinnedModelProcessor.ProcessMaterials(NodeContent node, ContentProcessorContext context) in D:\Desktop\2.0 Projects\Games\Block Games\2.0 Pipeline Extentions\Animation\Pipeline\SkinnedModelProcessor.cs:line 578
   at PipelineExtensionsAnimation.SkinnedModelProcessor.Process(NodeContent input, ContentProcessorContext in_Context) in D:\Desktop\2.0 Projects\Games\Block Games\2.0 Pipeline Extentions\Animation\Pipeline\SkinnedModelProcessor.cs:line 195
   at Microsoft.Xna.Framework.Content.Pipeline.ContentProcessor`2.Microsoft.Xna.Framework.Content.Pipeline.IContentProcessor.Process(Object input, ContentProcessorContext context)
   at MonoGame.Framework.Content.Pipeline.Builder.PipelineManager.ProcessContent(PipelineBuildEvent pipelineEvent)`

It doesn't help that I am quite sick today, I cant think strait... If you notice anything from this error list let me know.

**Comment by Jjagg** [comment]
> I have spent the last 5 hours trying to get a source version of the pipeline building (I also tried using your samplerparse branch)

So the issue is not just in this branch? If it is not caused by the changes in this PR you should open a new issue.

**Comment by KonajuGames** [comment]
@Danthekilla and @Jjagg, what is the status of this PR now?

**Comment by Jjagg** [comment]
I think it can be merged. Though the changes because of TinyPG version are annoying... It at least fixes the fact that samplers couldn't be used as parameters in functions. It doesn't fix the issue that samplers can't be used as local variables, but that's easier to work around and harder to fix. I'm not sure how to let the FX parser handle that, it should be aware that is is inside a function to solve the problem.

**Comment by Jjagg** [comment]
Though we should verify @Danthekilla's issue is not caused by this change. @Danthekilla is there a model you can share so we can test it.

**Comment by Danthekilla** [comment]
You should be right to merge it. I am pretty sure the problem went away, one of the other guys here was working on it I will see if he remembers. I don't believe that it was caused by this pull request however.

**Comment by KonajuGames** [comment]
I'd like to verify the TinyPG difference before merging.

**Comment by Jjagg** [comment]
@KonajuGames Fixed it. I used the release build of TinyPG and that doesn't print the comments.

**Comment by tomspilman** [comment]
@Jjagg - How does this affect https://github.com/MonoGame/MonoGame/pull/5354 ?  Which should I look to merge first?

**Comment by Jjagg** [comment]
They shouldn't conflict, so either is fine.

**Comment by KonajuGames** [comment]
That changeset looks a lot cleaner now.  I'll merge this now. Thanks @Jjagg

**Pull request merged**